docs/design/phases.md

```diff
@@ -0,0 +1,354 @@
+     - the hostname of the node
+       - I guess this might be a requested feature in opinionated setups, but might be a no-no in more advanced setups. Consensus here?
+     - the IPv4 address of the default route
+     - optional extra altnames that can be specified by the user
+ - a client certificate for the apiservers to connect to the kubelets securely (`apiserver-kubelet-client.crt`) using `ca.crt` as the CA with its private key (`apiserver-kubelet-client.key`). The certificate should:
+   - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+   - be in the `system:masters` organization
+ - a private key for signing ServiceAccount Tokens (`sa.key`) along with its public key (`sa.pub`)
+ - a CA for the front proxy (`front-proxy-ca.crt`) with its key (`front-proxy-ca.key`)
+ - a client cert for the front proxy client (`front-proxy-client.crt`) using `front-proxy-ca.crt` as the CA with its key (`front-proxy-client.key`)
+
+
+If a given certificate and private key pair both exist, the generation step will be skipped and those files will be validated and used for the prescribed use-case.
+This means the user can, for example, give his own CA, which then will be used for signing the rest of the certs.
+
+### Generate KubeConfig files on the master
+
+There should be:
+ - a KubeConfig file for kubeadm to use itself and the admin: `/etc/kubernetes/admin.conf`
+   - the "admin" here is defined as `kubeadm` itself and the actual person(s) that is administering the cluster and want to control the cluster
+     - with this file, the admin has full control over the cluster 
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - be in the `system:masters` organization
+ - a KubeConfig file for kubelet to use: `/etc/kubernetes/kubelet.conf`
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - be in the `system:nodes` organization
+     - have the name `system:node:<hostname>`
+ - a KubeConfig file for controller-manager: `/etc/kubernetes/controller-manager.conf`
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - have the name `system:kube-controller-manager`
+ - a KubeConfig file for scheduler: `/etc/kubernetes/scheduler.conf`
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - have the name `system:kube-scheduler`
+
+`ca.crt` is also embedded in the KubeConfig files.
+
+### Bootstrap the control plane
+
+#### etcd
+
+Determine if the user specified external etcd options. If not, etcd should:
```

Comment:
suggest: `kubeadm` determines if the user specified external etcd options in kubeadm config file.